### PR TITLE
qase-python-commons: add more debug logs

### PR DESCRIPTION
**Pull request by gibiw**
qase-python-commons: add more debug logs
--
Added a base class for all models. It helps to serialize the models to json.
Added logs before and after important functions in the common package.

---

qase-python: fixed an issue with xdist
--
Fixed the issue where the test run would not complete if the xDist package was disabled or not installed.

### CODE REVIEW

File: qase-python-commons/src/qase/commons/config.py [hunk 1]
```diff
@@ -59,3 +59,6 @@ def _get_config(self, key):
         for key in keys[:-1]:
             config = config.get(key, {})
         return config.get(keys[-1], None)
+
+    def to_json(self):
+        return json.dumps(self.config, indent=4)
```
Comment (NickVolynkin) [inline qase-python-commons/src/qase/commons/config.py:64]: Maybe it makes sense to add `sort_keys = True` to make the output more uniform and, thus, easier to read.

File: qase-python-commons/src/qase/commons/models/base.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import json
+
+from pathlib import PosixPath
+
+
+class Base:
+    def to_json(self) -> str:
+        return json.dumps(self, default=lambda o: o.__dict__ if hasattr(o, '__dict__') else str(o), indent=4)
```
Comment (NickVolynkin) [inline qase-python-commons/src/qase/commons/models/base.py:8]: Wouldn't it be easier to redefine `__str__` and then log just like

```diff
- self.logger.log_debug(f"Project {code} found: {response.result.to_json()}")
+ self.logger.log_debug(f"Project {code} found: {response.result}")
```

File: qase-python-commons/src/qase/commons/models/base.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import json
+
+from pathlib import PosixPath
+
+
+class Base:
```
Comment (NickVolynkin) [inline qase-python-commons/src/qase/commons/models/base.py:6]: Maybe, call it BaseModel?

Will we add or move more shared code to this class?

File: qase-python-commons/src/qase/commons/models/base.py [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+    def to_json(self) -> str:
```
Comment (NickVolynkin) [inline qase-python-commons/src/qase/commons/models/base.py:7]: Multiline logs are nice to read. However, they can break when we have parallel runs where multiple loggers write simultaneously.

File: qase-pytest/src/qase/pytest/plugin.py [hunk 5]
```diff
@@ -86,11 +86,12 @@ def pytest_sessionstart(self, session):
 
     def pytest_sessionfinish(self, session, exitstatus):
         if is_xdist_controller(session):
-            self.reporter.complete_run(exitstatus)
             QasePytestPlugin.drop_run_id()
         else:
             self.reporter.complete_worker()
 
+        self.reporter.complete_run(exitstatus)
```
Comment (NickVolynkin) [inline qase-pytest/src/qase/pytest/plugin.py:93]: xdist_controller is the thread that is spawning and managing other threads, right? If we don't check for it, won't we complete the run multiple times, from each thread?

Comment (NickVolynkin) [reply qase-pytest/src/qase/pytest/plugin.py:93]: And what about the case when the configuration is set to run.complete=False?

File: qase-python-commons/src/qase/commons/config.py (hunk 1 above)
Comment (gibiw) [reply qase-python-commons/src/qase/commons/config.py]: Done

File: qase-python-commons/src/qase/commons/models/base.py (hunk 2 above)
Comment (gibiw) [reply qase-python-commons/src/qase/commons/models/base.py]: Done

File: qase-pytest/src/qase/pytest/plugin.py (hunk 5 above)
Comment (gibiw) [reply qase-pytest/src/qase/pytest/plugin.py:93]: No, this method is called only once regardless of threads.
The `run.complete` variable is checked further in the code.

File: qase-python-commons/src/qase/commons/reporters/testops.py [hunk 6]
```diff
@@ -80,15 +81,18 @@ def _prepare_client(self) -> None:
...
             self.logger.log(f"Error at preparing API client: {e}", "error")
             raise ReporterException(e)
 
     # Method loads project from Qase TestOps by code and returns project data
     def _get_project(self, code: str) -> Dict:
         try:
+            self.logger.log_debug(f"Getting project {code}")
             response = ProjectsApi(self.client).get_project(code=code)
             if hasattr(response, 'result'):
+                self.logger.log_debug(f"Project {code} found: {response.result.to_json()}")
```
Comment (NickVolynkin) [inline qase-python-commons/src/qase/commons/reporters/testops.py:95]:
```suggestion
                self.logger.log_debug(f"Project {code} found: {response.result}")
```

File: qase-python-commons/src/qase/commons/reporters/testops.py [hunk 7]
```diff
@@ -99,12 +103,15 @@ def _get_project(self, code: str) -> Dict:
     # If environment not found or exception raised, returns None
     def _get_environment(self, environment: str, code: str) -> Union[str, None]:
         try:
+            self.logger.log_debug(f"Getting environment {environment}")
             api_instance = EnvironmentsApi(self.client)
             response = api_instance.get_environments(code=code)
             if hasattr(response, 'result') and hasattr(response.result, 'entities'):
                 for env in response.result.entities:
                     if env.slug == environment:
+                        self.logger.log_debug(f"Environment {environment} found: {env.to_json()}")
```
Comment (NickVolynkin) [inline qase-python-commons/src/qase/commons/reporters/testops.py:112]:
```suggestion
                        self.logger.log_debug(f"Environment {environment} found: {env}")
```

File: qase-python-commons/tests/tests_qase_commons/models/test_run.py [hunk 8]
```diff
@@ -59,5 +59,4 @@ def test_run():
     run.add_host_data(host_data)
     assert run.host_data == host_data
 
-    json_str = run.to_json()
-    assert isinstance(json.loads(json_str), dict)  # Ensure it's valid JSON and can be parsed
\ No newline at end of file
+    assert isinstance(json.loads(str(run)), dict)  # Ensure it's valid JSON and can be parsed
```
Comment (NickVolynkin) [inline qase-python-commons/tests/tests_qase_commons/models/test_run.py:62]: Please add a trailing new line

**Review by NickVolynkin — APPROVED** [review]

File: qase-python-commons/tests/tests_qase_commons/models/test_run.py (hunk 8 above)
Comment (gibiw) [reply qase-python-commons/tests/tests_qase_commons/models/test_run.py]: Fixed

**Pull request merged**